### PR TITLE
Add initial extensions handling via optional packages

### PR DESCRIPTION
**Pull request by twsouthwick**
## Summary of the Pull Request
This adds extension support via optional packages, including moving the Azure connector to such a package. This does not enable 3rd party plugins (optional packages run as the same identity and in the same process), but provides some infrastructure that could be helpful for 3rd part plugins as well.

## References
#555

## PR Checklist
* [ ] Closes #xxx
* [x] CLA signed. If not, go over [here](https://cla.opensource.microsoft.com/microsoft/Terminal) and sign the CLA
* [ ] Tests added/passed
* [ ] Requires documentation to be updated
* [ ] I've discussed this with core contributors already. If not checked, I'm ready to accept this work might be rejected in favor of a different grand plan. Issue number where discussion took place: #xxx

## Detailed Description of the Pull Request / Additional comments
This adds a few interfaces to allow composing of sources for terminal connections and moves the Azure connection source to its own library that is brought in via an optional package. The interfaces added are:

- ITerminalConnectionProvider: This can retrieve a connection factory by id and enumerate available factories
- ITerminalConnectionFactory: This provides information of an available connection as well as allows for creation of an instance
- TerminalConnectionStartupInfo: A class that contains information as to the initial start up state of a connection

A default `AppConnectionProvider` provides an implementation of `ITerminalConnectionProvider` that searches any optional packages and loads them via an exported function `GetConnectionProvider` that returns an `ITerminalConnectionProvider`. This could be expanded to include providers for all the default profiles so that nothing is hard coded but retrieved through a provider.

A caveat to loading optional packages is that the package must be signed by a valid certificate (can be a local dev), but without that it will fail. An error is written to the debug window to help diagnose that issue if it is encountered.

## Validation Steps Performted
I did a variety of manual testing. The plugin is only available when the package is deployed, so I tried various permutations of running without, with, then without. If the plugin is removed, I've just had it load the EchoConnection since I wasn't sure if there was a better way of handling errors but I didn't want it to crash.

Feel free to say no if you have other plans to enable this, but if there's interest I'll respond to feedback.

### CODE REVIEW

**Comment by twsouthwick** [comment]
Apparently the build is exhausting available disk space:

```
    C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\MSBuild\Microsoft\VisualStudio\v16.0\AppxPackage\Microsoft.AppXPackage.Targets(3501,5): error MSB3027: Could not copy "C:\Program Files (x86)\Microsoft SDKs\Windows Kits\10\ExtensionSDKs\Microsoft.VCLibs.Desktop\14.0\.\AppX\Retail\arm\Microsoft.VCLibs.arm.14.00.Desktop.appx" to "d:\a\1\s\pkg\appx\AppPackages\OpenConsolePackage_0.9.0.0_x86_Test\Dependencies\ARM\Microsoft.VCLibs.arm.14.00.Desktop.appx". Exceeded retry count of 10. Failed.  [d:\a\1\s\pkg\appx\OpenConsolePackage.wapproj]
         C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\MSBuild\Microsoft\VisualStudio\v16.0\AppxPackage\Microsoft.AppXPackage.Targets(3501,5): error MSB3021: Unable to copy file "C:\Program Files (x86)\Microsoft SDKs\Windows Kits\10\ExtensionSDKs\Microsoft.VCLibs.Desktop\14.0\.\AppX\Retail\arm\Microsoft.VCLibs.arm.14.00.Desktop.appx" to "d:\a\1\s\pkg\appx\AppPackages\OpenConsolePackage_0.9.0.0_x86_Test\Dependencies\ARM\Microsoft.VCLibs.arm.14.00.Desktop.appx". There is not enough space on the disk. [d:\a\1\s\pkg\appx\OpenConsolePackage.wapproj]
       C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\MSBuild\Microsoft\VisualStudio\v16.0\AppxPackage\Microsoft.AppXPackage.Targets(3501,5): error MSB3021:  [d:\a\1\s\pkg\appx\OpenConsolePackage.wapproj]
         C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\MSBuild\Microsoft\VisualStudio\v16.0\AppxPackage\Microsoft.AppXPackage.Targets(3501,5): error MSB3027: Could not copy "C:\Program Files (x86)\Microsoft SDKs\Windows Kits\10\ExtensionSDKs\Microsoft.VCLibs.Desktop\14.0\.\AppX\Retail\x64\Microsoft.VCLibs.x64.14.00.Desktop.appx" to "d:\a\1\s\pkg\appx\AppPackages\OpenConsolePackage_0.9.0.0_x86_Test\Dependencies\x64\Microsoft.VCLibs.x64.14.00.Desktop.appx". Exceeded retry count of 10. Failed.  [d:\a\1\s\pkg\appx\OpenConsolePackage.wapproj]
         C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\MSBuild\Microsoft\VisualStudio\v16.0\AppxPackage\Microsoft.AppXPackage.Targets(3501,5): error MSB3021: Unable to copy file "C:\Program Files (x86)\Microsoft SDKs\Windows Kits\10\ExtensionSDKs\Microsoft.VCLibs.Desktop\14.0\.\AppX\Retail\x64\Microsoft.VCLibs.x64.14.00.Desktop.appx" to "d:\a\1\s\pkg\appx\AppPackages\OpenConsolePackage_0.9.0.0_x86_Test\Dependencies\x64\Microsoft.VCLibs.x64.14.00.Desktop.appx". There is not enough space on the disk. [d:\a\1\s\pkg\appx\OpenConsolePackage.wapproj]
       C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\MSBuild\Microsoft\VisualStudio\v16.0\AppxPackage\Microsoft.AppXPackage.Targets(3501,5): error MSB3021:  [d:\a\1\s\pkg\appx\OpenConsolePackage.wapproj]
```

I'm not sure how to address that

**Comment by miniksa** [comment]
> Apparently the build is exhausting available disk space:
> 
> ```
>     C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\MSBuild\Microsoft\VisualStudio\v16.0\AppxPackage\Microsoft.AppXPackage.Targets(3501,5): error MSB3027: Could not copy "C:\Program Files (x86)\Microsoft SDKs\Windows Kits\10\ExtensionSDKs\Microsoft.VCLibs.Desktop\14.0\.\AppX\Retail\arm\Microsoft.VCLibs.arm.14.00.Desktop.appx" to "d:\a\1\s\pkg\appx\AppPackages\OpenConsolePackage_0.9.0.0_x86_Test\Dependencies\ARM\Microsoft.VCLibs.arm.14.00.Desktop.appx". Exceeded retry count of 10. Failed.  [d:\a\1\s\pkg\appx\OpenConsolePackage.wapproj]
>          C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\MSBuild\Microsoft\VisualStudio\v16.0\AppxPackage\Microsoft.AppXPackage.Targets(3501,5): error MSB3021: Unable to copy file "C:\Program Files (x86)\Microsoft SDKs\Windows Kits\10\ExtensionSDKs\Microsoft.VCLibs.Desktop\14.0\.\AppX\Retail\arm\Microsoft.VCLibs.arm.14.00.Desktop.appx" to "d:\a\1\s\pkg\appx\AppPackages\OpenConsolePackage_0.9.0.0_x86_Test\Dependencies\ARM\Microsoft.VCLibs.arm.14.00.Desktop.appx". There is not enough space on the disk. [d:\a\1\s\pkg\appx\OpenConsolePackage.wapproj]
>        C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\MSBuild\Microsoft\VisualStudio\v16.0\AppxPackage\Microsoft.AppXPackage.Targets(3501,5): error MSB3021:  [d:\a\1\s\pkg\appx\OpenConsolePackage.wapproj]
>          C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\MSBuild\Microsoft\VisualStudio\v16.0\AppxPackage\Microsoft.AppXPackage.Targets(3501,5): error MSB3027: Could not copy "C:\Program Files (x86)\Microsoft SDKs\Windows Kits\10\ExtensionSDKs\Microsoft.VCLibs.Desktop\14.0\.\AppX\Retail\x64\Microsoft.VCLibs.x64.14.00.Desktop.appx" to "d:\a\1\s\pkg\appx\AppPackages\OpenConsolePackage_0.9.0.0_x86_Test\Dependencies\x64\Microsoft.VCLibs.x64.14.00.Desktop.appx". Exceeded retry count of 10. Failed.  [d:\a\1\s\pkg\appx\OpenConsolePackage.wapproj]
>          C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\MSBuild\Microsoft\VisualStudio\v16.0\AppxPackage\Microsoft.AppXPackage.Targets(3501,5): error MSB3021: Unable to copy file "C:\Program Files (x86)\Microsoft SDKs\Windows Kits\10\ExtensionSDKs\Microsoft.VCLibs.Desktop\14.0\.\AppX\Retail\x64\Microsoft.VCLibs.x64.14.00.Desktop.appx" to "d:\a\1\s\pkg\appx\AppPackages\OpenConsolePackage_0.9.0.0_x86_Test\Dependencies\x64\Microsoft.VCLibs.x64.14.00.Desktop.appx". There is not enough space on the disk. [d:\a\1\s\pkg\appx\OpenConsolePackage.wapproj]
>        C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\MSBuild\Microsoft\VisualStudio\v16.0\AppxPackage\Microsoft.AppXPackage.Targets(3501,5): error MSB3021:  [d:\a\1\s\pkg\appx\OpenConsolePackage.wapproj]
> ```
> 
> I'm not sure how to address that

We're probably going to have to re-try doing the shared PCH thing I tried a week or two ago. It doesn't make the build any faster, but I think it saves on disk space.

The problem is that every WinRT project created here adds 1GB+ worth of obj/bin requirement. The hosted agents only have approximately 10GB of scratch space for obj/bin.

**Comment by zadjii-msft** [comment]
> We're probably going to have to re-try doing the shared PCH thing I tried a week or two ago. It doesn't make the build any faster, but I think it saves on disk space.

@miniksa I'm so sorry, I'm about to add one or two more cppwinrt projects D:

They're probably a week out though, so I'll wait to see how the shared pch thing goes first. If that works well enough, I'll hold them till that merges.

File: src/cascadia/TerminalApp/lib/AppConnectionProvider.cpp [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+                    if (!file.IsAvailable())
+                    {
+                        continue;
+                    }
+
+                    auto name = file.Name();
+
+                    if (IsPlugin(name))
+                    {
+                        auto dll = LoadPackagedLibrary(name.data(), 0);
```
Comment (zadjii-msft) [inline src/cascadia/TerminalApp/lib/AppConnectionProvider.cpp]: This is incredible that it was this easy

**Comment by miniksa** [comment]
> > We're probably going to have to re-try doing the shared PCH thing I tried a week or two ago. It doesn't make the build any faster, but I think it saves on disk space.
> 
> @miniksa Michael Niksa FTE I'm so sorry, I'm about to add one or two more cppwinrt projects D:
> 
> They're probably a week out though, so I'll wait to see how the shared pch thing goes first. If that works well enough, I'll hold them till that merges.

If it doesn't work, you're not going to be able to merge them at all.

File: src/cascadia/TerminalApp/lib/AppConnectionProvider.cpp (hunk 1 above)
Comment (zadjii-msft) [reply src/cascadia/TerminalApp/lib/AppConnectionProvider.cpp]: I mean, there's a lot of other stuff going on that's not easy, but I was under the impression that this specific bit would be the hard part. The rest is really just plumbing to make this bit work like magic.

**Comment by zadjii-msft** [comment]
So this is absolutely incredible. You've got a lot of great work that would certainly be really helpful for influencing our extension model in the future. Even though this is _great_, we're not really planning on shipping an extension model till after 1.0. I think we want to stabilize our internal models a bit more before we start having more developers take dependencies on them.

My major concern with this is the following: Would this work for 3rd-party plugins? From [this msdn link](https://docs.microsoft.com/en-us/windows/uwp/launch-resume/extend-your-app-with-services-extensions-packages) it seems like if we want to have 3rd parties publishing plugins for the Terminal, we'll _need_ to use an **App Extension**, which also _doesn't_ allow us to use "In-proc code plug-ins". So I'm not sure how this whole `LoadPackagedLibrary` thing would work with an app extension. Someone on an email thread last week mentioned maybe copying the extension's dll into the Shared directory, then loading that, though it seems super fragile, and maybe not the right solution.

So as great as this is, I think we need to also make sure that whatever solution we come up with is also able to support 3rd party developers.

CC'ing @Scottj1s @hassanuz (there are others interested as well)

**Comment by twsouthwick** [comment]
Yup, totally understand that this in itself doesn't help with 3rd party extensions and that extensions are probably not in scope for v1.0 (the optional packages were an easy way to show that the app doesn't need to know anything about the provider). The real value I'm hoping from this PR is some sort of abstraction of the terminal providers so that things *can* be plugged in easily without the app needing to know about it. I'm happy to remove the optional package stuff and have this focus more on that since some mechanism like what I put in here will probably be needed for any extension system.

I've started prototyping something for app service extensions, but nothing workable quite yet. I was thinking something along the lines of the following (using the infrastructure I've developed here):

1. The terminal itself declares itself to be an extension host
2. An application can then provide an extension for it
3. The terminal wraps calls to the appservice extension in the `ITerminalConnectionProvider` and handles the communication between the service and the terminal. The `GetFactories` method returns an array so that an extension could potentially provide multiple connection types.
4. The terminal handles connections via the `ITerminalConnectionProvider` and gets an `ITerminalConnections`, unaware that it is actually being proxied via an AppService.

This could potentially be used for any extension point; I'm just starting with `ITerminalConnection` because that has the most actual use (as seen by things like the Azure connector). It seems like most of the `ITerminalConnection` stuff is fine being asynchronous in a fire and forget way; some of the other extensions I've seen discussed may need to actually use async methods correctly to get the response. I'm explicitly not worrying about that right now as async doesn't seem to be used in the code base and didn't want to through that in as well.

**Comment by zadjii-msft** [comment]
I've been looking into that route as well, but part of the reason I was avoiding it is I believe that we'll eventually want extensions that host UI elements to also be able to run in the terminal. While the app service connection would certainly work fine for an `ITerminalConnection`, I'm not sure how it'd work for some of [these other scenarios](https://github.com/microsoft/terminal/issues?q=is%3Aissue+is%3Aopen+label%3AArea-Extensibility).

**Comment by miniksa** [comment]
You should be able to merge after #2271 is complete and this should work.

File: src/cascadia/TerminalApp/lib/AppConnectionProvider.cpp [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+
+                            if (procAddress == nullptr)
+                            {
+                                DebugOut(L"Could not find connection provider via entrypoint '%S' in '%s' from '%s'. Must have an entrypoint named '%S'.\n", ConnectionProviderName, name.data(), package.Id().FullName().data());
+                                FreeLibrary(dll);
+                            }
+                            else
+                            {
+                                DebugOut(L"Found connection provider in '%s' from '%s'.\n", name.data(), package.Id().FullName().data());
+                                auto func = (GetConnectionProvider)procAddress;
```
Comment (dlong11) [inline src/cascadia/TerminalApp/lib/AppConnectionProvider.cpp]: nit: prefer a named cast instead a c style cast

File: src/cascadia/TerminalConnection/TerminalConnectionStartupInfo.h [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+        {
+            return _startingDirectory;
+        }
+
+        void StartingDirectory(hstring startingDirectory)
+        {
+            _startingDirectory = startingDirectory;
+        }
+
+        unsigned int InitialRows()
```
Comment (dlong11) [inline src/cascadia/TerminalConnection/TerminalConnectionStartupInfo.h]: Any reason these are not ints?

**Comment by zadjii-msft** [comment]
Hey so this PR has been open for a while - in the interest of housekeeping, I'm going to close it. It was an **amazing** proof-of-concept, but I think we're going to need to go another route in order to support third-party extensions. 

Thanks again for the contribution!

**Pull request closed without merging**